internal/server/recon/place_name.go

```diff
@@ -0,0 +1,271 @@
+	for info := range placeInfoSet {
+		place := &pb.ResolvePlaceNamesResponse_Place{
+			Name: info.placeName,
+			Type: info.placeType,
+		}
+
+		if placeIDs, ok := nameToPlaceIDs.Load(info.placeName); ok {
+			allDCIDs := []string{}
+			for _, placeID := range placeIDs.([]string) {
+				if dcids, ok := placeIDToDCIDs[placeID]; ok {
+					allDCIDs = append(allDCIDs, dcids...)
+				}
+			}
+			sort.Strings(allDCIDs)
+
+			if len(allDCIDs) != 0 {
+				if info.placeType == "" {
+					// No type filtering.
+					place.Dcids = allDCIDs
+				} else {
+					// Type filtering.
+					filteredDCIDs := []string{}
+					for _, dcid := range allDCIDs {
+						typeSet, ok := dcidToTypeSet[dcid]
+						if !ok {
+							continue
+						}
+						if _, ok := typeSet[info.placeType]; ok {
+							filteredDCIDs = append(filteredDCIDs, dcid)
+						}
+					}
+					place.Dcids = filteredDCIDs
+				}
+			}
+		}
+
+		resp.Places = append(resp.Places, place)
+	}
+
+	return resp, nil
+}
+
+func resolvePlaceIDs(
+	ctx context.Context,
+	mapsClient *maps.Client,
+	nameSet map[string]struct{},
+) (*sync.Map /* name -> place IDs */, *sync.Map /* place ID set */, error) {
+	type placeInfo struct {
+		name     string
+		placeIDs []string
+	}
+	var wg sync.WaitGroup
+
+	// Send place names to channel.
+	nameChan := make(chan string)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		defer close(nameChan)
+		for name := range nameSet {
+			nameChan <- name
+		}
+	}()
+
+	// Receive and process resolved place IDs.
+	nameToPlaceIDs := &sync.Map{}
+	placeIDSet := &sync.Map{}
+	placeInfoChan := make(chan placeInfo, maxMapsAPICallsInParallel)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for placeInfo := range placeInfoChan {
+			nameToPlaceIDs.Store(placeInfo.name, placeInfo.placeIDs)
+			for _, placeID := range placeInfo.placeIDs {
+				placeIDSet.Store(placeID, struct{}{})
+			}
+		}
+	}()
+
+	// Call Maps API to find place IDs in parallel.
+	// The errors in the Goroutines need to be captured, so we use errgroup.
+	eg, errCtx := errgroup.WithContext(ctx)
+	for i := 0; i < maxMapsAPICallsInParallel; i++ {
+		eg.Go(func() error {
+			for name := range nameChan {
+				placeIDs, err := findPlaceIDsFromName(errCtx, mapsClient, name)
+				if err != nil {
+					return err
+				}
+				placeInfoChan <- placeInfo{name: name, placeIDs: placeIDs}
+			}
+			return nil
+		})
+	}
+	if err := eg.Wait(); err != nil {
+		return nil, nil, err
+	}
+	close(placeInfoChan)
+
+	wg.Wait()
+
+	return nameToPlaceIDs, placeIDSet, nil
+}
+
+func findPlaceIDsFromName(
+	ctx context.Context,
+	mapsClient *maps.Client,
+	name string,
+) ([]string, error) {
+	resp, err := mapsClient.FindPlaceFromText(ctx, &maps.FindPlaceFromTextRequest{
+		Input:     name,
+		InputType: maps.FindPlaceFromTextInputTypeTextQuery,
+		Fields:    []maps.PlaceSearchFieldMask{maps.PlaceSearchFieldMaskPlaceID},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	placeIDs := []string{}
+	for _, candidate := range resp.Candidates {
```

Comment:
If this is the candidate type - https://github.com/googlemaps/google-maps-services-go/blob/391036138cfed529e87aaad49922e96da4a393da/places.go#L331-L332 - looks like it might also include the type.  So in future, we could consider using that to prune unexpected types.